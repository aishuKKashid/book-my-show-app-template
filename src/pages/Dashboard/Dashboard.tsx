import { PageContent } from '@components/layout';
import { Button, Input, Modal, Text } from '@medly-components/core';
import axios from 'axios';
import React, { FC, useEffect, useState } from 'react';
import { DashboardProps } from './types';

export const Dashboard: FC<DashboardProps> = ({ isLoading }) => {
    const [movies, setMovies] = useState([]);
    const [state, setState] = useState(false);
    const [movie, setMovie] = useState({
        title: '',
        duration: 0
    });

    const getMovies = async () => {
        console.log('In getMovies');
        await axios
            .get(`/api/movies`)
            .then(res => setMovies(res.data))
            .catch(err => console.log(err));
    };

    useEffect(() => {
        getMovies();
    }, []);

    const openModal = () => setState(true);
    const closeModal = () => setState(false);

    const addTitle = (e: any) => {
        const { value } = e.target;
        setMovie({ ...movie, title: value });
    };

    const addDuration = (e: any) => {
        const { value } = e.target;
        setMovie({ ...movie, duration: value });
    };

    const addMovie = (values: any) => {
        // if()
        axios
            .post(`/api/movies`, values)
            .then(res => {
                console.log(res.data);
                closeModal();
                getMovies();
                return null;
            })
            .catch(err => console.log(err));
    };
    console.log(movie);
    return (
        <>
            <PageContent isLoading={isLoading}>
                <Text textWeight="Strong" textVariant="body1">
                    List of Movies
                </Text>
                <Button variant="outlined" style={styles.add} onClick={openModal}>
                    Add a Movie
                </Button>
                <MovieList movieList={movies} />
            </PageContent>
            <Modal open={state} onCloseModal={closeModal} data-testid="modal">
                <Modal.Header>Add Movie</Modal.Header>
                <Modal.Content>
                    <Input
                        id="name-input"
                        type="text"
                        fullWidth
                        label="Title"
                        placeholder="Enter Movie Title"
                        name="title"
                        onChange={e => addTitle(e)}
                        required
                    />
                    <Input
                        id="number-input"
                        type="number"
                        fullWidth
                        label="Duration"
                        placeholder="Enter Duration"
                        onChange={e => addDuration(e)}
                        required
                    />
                </Modal.Content>
                <Modal.Actions>
                    <Button variant="outlined" onClick={() => addMovie(movie)}>
                        Add Movie
                    </Button>
                </Modal.Actions>
            </Modal>
        </>
    );
};
Dashboard.displayName = 'Dashboard';

export const MovieList = ({ movieList }: any) => {
    console.log(movieList);
    return (
        <div>
            {movieList?.map((movie: any) => {
                return (
                    <div key={movie?.id} data-testid="movie">
                        <Text style={styles.title}>{movie?.title}</Text>
                        <Text style={styles.duration}>{movie?.duration}</Text>
                    </div>
                );
            })}
        </div>
    );
};

const styles = {
    title: {
        padding: 8
    },
    duration: {
        padding: 8
    },
    add: {
        margin: 16
    }
};
