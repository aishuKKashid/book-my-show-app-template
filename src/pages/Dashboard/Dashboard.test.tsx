import { mockAxios, render } from '@test-utils';
import { fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Dashboard from './';

describe('Dashboard', () => {
    beforeEach(async () =>
        mockAxios.onGet(`/api/movies`).reply(200, [
            {
                id: 1,
                title: 'Avengers',
                duration: 8111
            },
            { id: 2, title: 'Dr. Strange', duration: 360 }
        ])
    );

    afterEach(() => mockAxios.resetHistory());

    it('should render properly', async () => {
        const { container } = render(<Dashboard />);
        // await waitFor(() => expect(mockAxios.history.get.length).toEqual(1));
        expect(container).toMatchSnapshot();
    });

    it('should show loading if isLoading Prop is true', () => {
        const { container } = render(<Dashboard isLoading />);
        expect(container).toMatchSnapshot();
    });

    it('should render button properly', () => {
        render(<Dashboard />);
        const button = screen.getByRole('button', { name: /Add a Movie/i });
        expect(button).toBeInTheDocument();
        expect(button).toHaveTextContent(/Add a Movie/);
    });

    it('should show list', async () => {
        render(<Dashboard />);
        const a = await screen.findByText(/Avengers/i);
        await screen.findByText(/Dr. Strange/i);
        expect(a).toBeInTheDocument();
    });

    it('check if the modal is getting render on click', () => {
        render(<Dashboard />);
        const button = screen.getByText(/Add a Movie/i);
        fireEvent.click(button);
        expect(screen.getByTestId(/modal/i)).toBeInTheDocument();
    });

    it('check if movie is getting added', async () => {
        // mockAxios.onPost('/api/movies', { title: 'ABCD', duration: 420 }).reply(200, 3);
        // mockAxios.onGet('/api/movies').reply(200, [{ id: 3, title: 'ABCD', duration: 420 }]);
        render(<Dashboard />);
        const button = screen.getByText(/Add a Movie/i);
        fireEvent.click(button);

        // const titleInput = screen.getAllByLabelText(/Title/i);
        // userEvent.type(titleInput, ['ABCD']);
        // expect(titleInput).toHaveValue('ABCD');

        const titleInput = screen.getByTestId(/name-input/i);
        userEvent.type(titleInput, 'ABCD');
        expect(titleInput).toHaveValue('ABCD');

        const durationInput = screen.getByTestId(/number-input/i);
        userEvent.type(durationInput, '300');
        expect(titleInput).toHaveValue('300');

        const a = await screen.findByText(/ABCD/i);
        expect(screen.getByTestId(/modal/i)).toBeInTheDocument();
        expect(a).toBeInTheDocument();
    });
});
