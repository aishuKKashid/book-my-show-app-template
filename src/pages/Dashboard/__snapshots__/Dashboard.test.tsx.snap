// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render properly 1`] = `
.c1 {
  margin: 0;
  color: inherit;
  font-size: 1.6rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.6rem;
  text-align: initial;
  display: inline-block;
}

.c3 {
  margin: 0;
  color: inherit;
  font-size: 1.6rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.6rem;
  text-align: initial;
  display: inline-block;
}

.c2 {
  border: none;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  padding: 1rem 2.4rem 1.2rem;
  border-radius: 0.8rem;
  background-color: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2:hover {
  cursor: pointer;
}

.c2:focus {
  outline: none;
}

.c2:disabled {
  cursor: not-allowed;
}

.c2::after {
  content: '';
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: transparent;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  border-radius: 0.8rem;
  border-style: solid;
}

.c2:disabled {
  color: #98A7B7;
  background-color: transparent;
}

.c2:disabled::after {
  border-width: 0.1rem;
  border-color: #98A7B7;
}

.c2:not(:disabled):not(:hover) {
  color: #3872D2;
  background-color: transparent;
}

.c2:not(:disabled):not(:hover)::after {
  border-width: 0.1rem;
  border-color: #3872D2;
}

.c2:not(:disabled):active {
  color: #275093;
  background-color: #EBF1FA;
}

.c2:not(:disabled):active::after {
  border-width: 0.2rem;
  border-color: #275093;
}

.c2:not(:disabled):not(:active):hover {
  color: #3061B3;
  background-color: transparent;
  box-shadow: 0 0.2rem 0.8rem rgba(48,97,179,0.35);
}

.c2:not(:disabled):not(:active):hover::after {
  border-width: 0.2rem;
  border-color: #3061B3;
}

.c0 {
  background-color: #eff2f4;
  overflow: auto;
  padding-top: 32px;
}

@media (max-width:1279px) {
  .c0 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:1280px) {
  .c0 {
    padding-left: calc(24px + (((100vw - 1280px) / 160) * 16));
    padding-right: calc(24px + (((100vw - 1280px) / 160) * 16));
  }
}

@media (min-width:1440px) {
  .c0 {
    padding-left: 40px;
    padding-right: 40px;
  }
}

<div>
  <main
    class="c0"
  >
    <strong
      class="c1"
    >
      List of Movies
    </strong>
    <button
      class="c2"
      style="margin: 16px;"
      type="button"
    >
      <span
        class="c3"
      >
        Add a Movie
      </span>
    </button>
    <div />
  </main>
</div>
`;

exports[`Dashboard should show loading if isLoading Prop is true 1`] = `
.c3 {
  margin: 0;
  color: inherit;
  font-size: 1.6rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.6rem;
  text-align: initial;
  display: inline-block;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.6rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.6rem;
  text-align: initial;
  display: inline-block;
}

.c2 {
  font-size: 56px;
}

.c2 * {
  fill: #3872D2;
}

.c4 {
  border: none;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  padding: 1rem 2.4rem 1.2rem;
  border-radius: 0.8rem;
  background-color: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4:hover {
  cursor: pointer;
}

.c4:focus {
  outline: none;
}

.c4:disabled {
  cursor: not-allowed;
}

.c4::after {
  content: '';
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: transparent;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  border-radius: 0.8rem;
  border-style: solid;
}

.c4:disabled {
  color: #98A7B7;
  background-color: transparent;
}

.c4:disabled::after {
  border-width: 0.1rem;
  border-color: #98A7B7;
}

.c4:not(:disabled):not(:hover) {
  color: #3872D2;
  background-color: transparent;
}

.c4:not(:disabled):not(:hover)::after {
  border-width: 0.1rem;
  border-color: #3872D2;
}

.c4:not(:disabled):active {
  color: #275093;
  background-color: #EBF1FA;
}

.c4:not(:disabled):active::after {
  border-width: 0.2rem;
  border-color: #275093;
}

.c4:not(:disabled):not(:active):hover {
  color: #3061B3;
  background-color: transparent;
  box-shadow: 0 0.2rem 0.8rem rgba(48,97,179,0.35);
}

.c4:not(:disabled):not(:active):hover::after {
  border-width: 0.2rem;
  border-color: #3061B3;
}

.c0 {
  background-color: #eff2f4;
  overflow: auto;
  padding-top: 32px;
}

.c1 {
  position: fixed;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  background-color: rgba(255,255,255,0.7);
  z-index: 1;
}

@media (max-width:1279px) {
  .c0 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media (min-width:1280px) {
  .c0 {
    padding-left: calc(24px + (((100vw - 1280px) / 160) * 16));
    padding-right: calc(24px + (((100vw - 1280px) / 160) * 16));
  }
}

@media (min-width:1440px) {
  .c0 {
    padding-left: 40px;
    padding-right: 40px;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <svg
        class="c2"
        height="1em"
        viewBox="0 0 100 100"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M82 35.7C74.1 18 53.4 10.1 35.7 18S10.1 46.6 18 64.3l7.6-3.4c-6-13.5 0-29.3 13.5-35.3s29.3 0 35.3 13.5l7.6-3.4z"
        >
          <animatetransform
            attributeName="transform"
            attributeType="XML"
            dur="2s"
            from="0 50 50"
            repeatCount="indefinite"
            to="360 50 50"
            type="rotate"
          />
        </path>
      </svg>
    </div>
    <strong
      class="c3"
    >
      List of Movies
    </strong>
    <button
      class="c4"
      style="margin: 16px;"
      type="button"
    >
      <span
        class="c5"
      >
        Add a Movie
      </span>
    </button>
    <div />
  </main>
</div>
`;
